
# Program to handle missing data in Pandas Dataframe

In [1]:
import pandas as pd
import numpy as np

### The *Fat* Data consists of the amount of fat that has been accumulated in the Chest, Abdomen, Hip, Thigh, Knee and Forearm which is influencing the weight of an individual.

In [2]:
df = pd.read_csv("fat.txt", delimiter="\t", header=0)
df

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
0    154.25   93.1     85.2   94.5   59.0  37.3     27.4
1    173.25   93.6     83.0   98.7   58.7  37.3     28.9
2    154.00   95.8     87.9   99.2   59.6  38.9     25.2
3    184.75  101.8     86.4  101.2   60.1  37.3     29.4
4    184.25   97.3    100.0  101.9   63.2  42.2     27.7
..      ...    ...      ...    ...    ...   ...      ...
245  134.25   89.2     83.6   88.8   49.6  34.8     25.7
246  201.00  108.5    105.0  104.5   59.6  40.8     28.6
247  186.75  111.1    111.5  101.7   60.3  37.3     27.2
248  190.75  108.3    101.3   97.8   56.0  41.6     29.4
249  207.50  112.4    108.5  107.1   59.3  42.2      NaN

[250 rows x 7 columns]

In [3]:
df.describe()

Weight       Chest     Abdomen         Hip       Thigh        Knee  \
count  250.000000  248.000000  247.000000  248.000000  247.000000  249.000000   
mean   178.083200  100.679435   92.214980   99.652419   59.231579   38.537349   
std     27.035488    8.169402   10.222054    6.468897    4.908278    2.317676   
min    118.500000   79.300000   69.400000   85.000000   47.200000   33.000000   
25%    158.500000   94.350000   84.450000   95.475000   56.000000   36.900000   
50%    176.125000   99.600000   90.900000   99.300000   59.000000   38.500000   
75%    196.750000  105.300000   99.000000  103.275000   62.200000   39.900000   
max    262.750000  128.300000  126.200000  125.600000   74.400000   46.000000   

          Forearm  
count  245.000000  
mean    28.674286  
std      2.032914  
min     21.000000  
25%     27.300000  
50%     28.700000  
75%     30.000000  
max     34.900000

###  To Display the top 5 rows of the data

In [4]:
df.head()

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
0  154.25   93.1     85.2   94.5   59.0  37.3     27.4
1  173.25   93.6     83.0   98.7   58.7  37.3     28.9
2  154.00   95.8     87.9   99.2   59.6  38.9     25.2
3  184.75  101.8     86.4  101.2   60.1  37.3     29.4
4  184.25   97.3    100.0  101.9   63.2  42.2     27.7

### To check if the data has missing values we can either use the **info()** function of pandas which results in the number of not null records or use the isnull() function

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Weight   250 non-null    float64
 1   Chest    248 non-null    float64
 2   Abdomen  247 non-null    float64
 3   Hip      248 non-null    float64
 4   Thigh    247 non-null    float64
 5   Knee     249 non-null    float64
 6   Forearm  245 non-null    float64
dtypes: float64(7)
memory usage: 13.8 KB


### We can see that our data consists of 250 values of each attribute, but chest,abdomen,ip, thigh,knee and forearm do not contain 250 values. Therefore we can say that, there are some missing values in the data

In [6]:
print(df.isnull().sum())

Weight     0
Chest      2
Abdomen    3
Hip        2
Thigh      3
Knee       1
Forearm    5
dtype: int64


### The isnull() gives a boolean value, We have combined it with sum() function so that we get the count of missing values in each columns.

### Now we have various ways to deal with this missing values:
#### 1. dropna()
#### 2. fillna()
#### 3. replace()


## Deleting columns with missing data

* If we have a very small amount of data, and the column containing missing values doesn't impact the final result or its dependent variables, then we can delete the column using **dataframe.dropna(axis =1, how ="any")** which deletes any column that contains null values or **dataframe.drop(["Column_Name"], axis =1)**.
***************
* In our data, the weight of an individual is dependent on the fat contains in given parts of the body, deleting any column would affect the accuracy of the weight value. Therefore, we try to delete rows which contains least missing values and try to fill in other missing values.

###  First let us display the records with missing values

In [7]:
df[df.isnull().any(axis=1)]

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
32   168.00    NaN     79.1   94.5   57.3  36.2     30.0
40   217.00  113.3    111.2  114.1    NaN  40.9     29.8
61   178.00  105.3      NaN   99.7   60.8  40.1     29.0
62   205.50  105.3      NaN  108.3   65.0  41.2     31.1
69   146.75   88.5     82.8   95.5   58.9  37.6      NaN
70   160.75   93.6     82.9   96.3   52.9  37.5      NaN
71   125.00   87.7     76.0   88.6   50.9  35.4      NaN
107  168.50   97.8     92.3    NaN   57.5  36.8     26.0
141  159.75   92.1     77.1   93.9    NaN  36.1     27.2
226  174.50    NaN      NaN    NaN    NaN   NaN      NaN
249  207.50  112.4    108.5  107.1   59.3  42.2      NaN

### Deleting rows with missing data

In [8]:
display(df.iloc[226])

Weight     174.5
Chest        NaN
Abdomen      NaN
Hip          NaN
Thigh        NaN
Knee         NaN
Forearm      NaN
Name: 226, dtype: float64

In [9]:
df.dropna(subset =["Knee"], axis = 0, inplace =True)


### Verification

In [10]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 249 entries, 0 to 249
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Weight   249 non-null    float64
 1   Chest    248 non-null    float64
 2   Abdomen  247 non-null    float64
 3   Hip      248 non-null    float64
 4   Thigh    247 non-null    float64
 5   Knee     249 non-null    float64
 6   Forearm  245 non-null    float64
dtypes: float64(7)
memory usage: 15.6 KB
None


#### Now by using the dropna() function we deleted the record of Knee that contained missing value and verified it through the info() function which displayed only 249 values.

### Filling the Missing Values

### Filling with a random approximate value
### Here I have filled the NaN value of Chest with 64 as it was the approximate value.

In [11]:
display(df.iloc[[32,226]])

Weight  Chest  Abdomen   Hip  Thigh  Knee  Forearm
32   168.00    NaN     79.1  94.5   57.3  36.2     30.0
227  167.75   98.6     93.0  97.0   55.4  38.8     29.7

In [12]:
df["Chest"] = df["Chest"].fillna(64)

### Verification

In [13]:
display(df.iloc[[32,33,34]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
32  168.00   64.0     79.1   94.5   57.3  36.2     30.0
33  218.50  111.7    100.5  108.3   67.1  44.2     31.5
34  247.25  117.0    115.6  116.1   71.2  43.3     31.7

### Forward or Backward Fill
### Filling in with the previous or succeeding values of a column. 
###  The missing values of Abdomen have been filled with the previous abdomen values.

In [14]:
display(df.iloc[[60,61,62,63]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
60  193.25  107.6    102.4   99.4   61.0  39.4     29.9
61  178.00  105.3      NaN   99.7   60.8  40.1     29.0
62  205.50  105.3      NaN  108.3   65.0  41.2     31.1
63  183.50  103.0    100.3  104.2   64.8  40.2     30.1

In [15]:
df["Abdomen"] = df["Abdomen"].fillna(method = "bfill")

### verification

In [16]:
display(df.iloc[[60,61,62,63,]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
60  193.25  107.6    102.4   99.4   61.0  39.4     29.9
61  178.00  105.3    100.3   99.7   60.8  40.1     29.0
62  205.50  105.3    100.3  108.3   65.0  41.2     31.1
63  183.50  103.0    100.3  104.2   64.8  40.2     30.1

### Using Replace()

### The missing value of thigh has been replaced using replace function

In [17]:
display(df.iloc[[40,141]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
40   217.00  113.3    111.2  114.1    NaN  40.9     29.8
141  159.75   92.1     77.1   93.9    NaN  36.1     27.2

In [18]:
df["Thigh"] = df["Thigh"].replace(np.nan, value= 57)

### Verification

In [19]:
display(df.iloc[[40,141]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
40   217.00  113.3    111.2  114.1   57.0  40.9     29.8
141  159.75   92.1     77.1   93.9   57.0  36.1     27.2

### Filling the missing values with some statistical measures  

### Mean

In [20]:
display(df.iloc[[69,70,71,248]])

Weight  Chest  Abdomen    Hip  Thigh  Knee  Forearm
69   146.75   88.5     82.8   95.5   58.9  37.6      NaN
70   160.75   93.6     82.9   96.3   52.9  37.5      NaN
71   125.00   87.7     76.0   88.6   50.9  35.4      NaN
249  207.50  112.4    108.5  107.1   59.3  42.2      NaN

In [21]:

df1_mean = df["Forearm"].fillna(df["Forearm"].mean()).copy(deep=False)
display(df1_mean.iloc[[69,70,71,248]])

69     28.674286
70     28.674286
71     28.674286
249    28.674286
Name: Forearm, dtype: float64

### Median

In [22]:
df2_med = df["Forearm"].fillna(df["Forearm"].median()).copy(deep=False)
df2_med.iloc[[68,69,70,71,72]]

68    27.3
69    28.7
70    28.7
71    28.7
72    27.0
Name: Forearm, dtype: float64

###  Filling null values of Forearm column with mean values in the origial dataframe

In [23]:
df["Forearm"].fillna(df["Forearm"].mean(), inplace =True)
df.iloc[[68,69,70,71,72]]


Weight  Chest  Abdomen   Hip  Thigh  Knee    Forearm
68  167.50   97.6     91.5  98.5   56.6  38.6  27.300000
69  146.75   88.5     82.8  95.5   58.9  37.6  28.674286
70  160.75   93.6     82.9  96.3   52.9  37.5  28.674286
71  125.00   87.7     76.0  88.6   50.9  35.4  28.674286
72  143.00   93.4     83.3  93.0   55.5  35.2  27.000000

### Filling null values of Forearm column with median values in the origial dataframe

In [24]:
display(df.iloc[107])

Weight     168.5
Chest       97.8
Abdomen     92.3
Hip          NaN
Thigh       57.5
Knee        36.8
Forearm     26.0
Name: 107, dtype: float64

In [25]:
df["Hip"].fillna(df["Hip"].median(), inplace =True)
df.iloc[[107]]

Weight  Chest  Abdomen   Hip  Thigh  Knee  Forearm
107   168.5   97.8     92.3  99.3   57.5  36.8     26.0

### Finally our dataframe is free from null values

In [26]:
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [Weight, Chest, Abdomen, Hip, Thigh, Knee, Forearm]
Index: []

In [37]:
df.iloc[[32,40,41,61,68,70,71,107,141,248]]

Weight  Chest  Abdomen    Hip  Thigh  Knee    Forearm
32   168.00   64.0     79.1   94.5   57.3  36.2  30.000000
40   217.00  113.3    111.2  114.1   57.0  40.9  29.800000
41   212.00  106.6    104.3  106.0   65.0  40.2  31.500000
61   178.00  105.3    100.3   99.7   60.8  40.1  29.000000
68   167.50   97.6     91.5   98.5   56.6  38.6  27.300000
70   160.75   93.6     82.9   96.3   52.9  37.5  28.674286
71   125.00   87.7     76.0   88.6   50.9  35.4  28.674286
107  168.50   97.8     92.3   99.3   57.5  36.8  26.000000
141  159.75   92.1     77.1   93.9   57.0  36.1  27.200000
249  207.50  112.4    108.5  107.1   59.3  42.2  28.674286

In [40]:
print(df.isnull().sum())

Weight     0
Chest      0
Abdomen    0
Hip        0
Thigh      0
Knee       0
Forearm    0
dtype: int64


In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 249 entries, 0 to 249
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Weight   249 non-null    float64
 1   Chest    249 non-null    float64
 2   Abdomen  249 non-null    float64
 3   Hip      249 non-null    float64
 4   Thigh    249 non-null    float64
 5   Knee     249 non-null    float64
 6   Forearm  249 non-null    float64
dtypes: float64(7)
memory usage: 15.6 KB
